File: packages/cpp/examples/runtime_adapter.cpp
#include "axllm/axllm.hpp"
#include <iostream>

struct DemoSession : axllm::AxCodeSession {
  axllm::Value globals;
  axllm::Value create_options;
  bool closed = false;

  DemoSession(axllm::Value globals_, axllm::Value options_) : globals(std::move(globals_)), create_options(std::move(options_)) {}

  axllm::Value execute(axllm::Value code, axllm::Value options = axllm::Value::object()) override {
    if (!axllm::Core::truthy(axllm::Core::map_contains(options, "reservedNames"))) throw axllm::AxError("fixture", "missing reservedNames");
    if (axllm::equal(code, "timeout()")) return axllm::RuntimeEnvelope::timeout("demo timeout");
    axllm::Core::set(globals, "answer", "runtime");
    return axllm::RuntimeEnvelope::final_payload({axllm::object({{"answer", axllm::Core::get(globals, "answer")}})});
  }

  axllm::Value inspect(axllm::Value = axllm::Value::object()) override { return globals; }
  axllm::Value snapshot_globals(axllm::Value = axllm::Value::object()) override {
    return axllm::object({{"version", 1}, {"bindings", globals}, {"globals", globals}, {"closed", closed}});
  }
  axllm::Value patch_globals(axllm::Value snapshot, axllm::Value options = axllm::Value::object()) override {
    globals = axllm::Core::get(snapshot, "bindings", axllm::Value::object());
    return snapshot_globals(options);
  }
  axllm::Value close() override {
    closed = true;
    return axllm::object({{"closed", true}});
  }
};

struct DemoRuntime : axllm::AxCodeRuntime {
  axllm::RuntimeCapabilities capabilities;
  std::vector<std::unique_ptr<DemoSession>> sessions;

  DemoRuntime() {
    capabilities.language = "Python";
    capabilities.snapshot = true;
    capabilities.patch = true;
  }

  std::string language() const override { return "Python"; }
  axllm::AxCodeSession* create_session(axllm::Value globals, axllm::Value options = axllm::Value::object()) override {
    sessions.push_back(std::make_unique<DemoSession>(std::move(globals), std::move(options)));
    return sessions.back().get();
  }
};

int main() {
  DemoRuntime runtime;
  auto qa = axllm::agent("question:string -> answer:string", axllm::object({{"runtime", axllm::object({{"language", "Python"}})}}));
  axllm::Value out = qa.test(runtime, "final()", axllm::object({{"question", "adapter"}}));
  if (!axllm::equal(axllm::Core::get(out, "kind"), "final")) return 1;
  if (!runtime.sessions.back()->closed) return 2;

  auto runner = axllm::agent("question:string -> answer:string", axllm::object({{"runtime", axllm::object({{"language", "Python"}})}}));
  axllm::Value step = runner.execute_actor_step(runtime, "final()", axllm::object({{"question", "adapter"}}));
  if (!axllm::equal(axllm::Core::get(step, "kind"), "final")) return 3;
  axllm::Value snapshot = runner.export_session_state();
  runner.restore_session_state(snapshot);
  axllm::Value timeout = runner.execute_actor_step(runtime, "timeout()", axllm::object({{"question", "adapter"}}));
  if (!axllm::equal(axllm::Core::get(timeout, "error_category"), "timeout")) return 4;
  std::cout << "cpp-runtime-adapter-ok\n";
}

File: packages/python/examples/runtime_adapter.py
from axllm import AxCodeRuntime, AxCodeSession, RuntimeCapabilities, RuntimeEnvelope, agent


class DemoSession(AxCodeSession):
    def __init__(self, globals, options=None):
        self.globals = dict(globals or {})
        self.create_options = dict(options or {})
        self.closed = False

    def execute(self, code, options=None):
        assert "reservedNames" in (options or {}), options
        if code == "timeout()":
            return RuntimeEnvelope.timeout("demo timeout")
        self.globals["answer"] = "runtime"
        return RuntimeEnvelope.final({"answer": self.globals["answer"]})

    def inspect_globals(self, options=None):
        return dict(self.globals)

    def snapshot_globals(self, options=None):
        return {"version": 1, "bindings": dict(self.globals), "globals": dict(self.globals), "closed": self.closed}

    def patch_globals(self, snapshot, options=None):
        self.globals = dict((snapshot or {}).get("bindings") or {})
        return self.snapshot_globals(options)

    def close(self):
        self.closed = True
        return {"closed": True}


class DemoRuntime(AxCodeRuntime):
    language = "Python"

    def __init__(self):
        self.capabilities = RuntimeCapabilities(language="Python", snapshot=True, patch=True).to_dict()
        self.sessions = []

    def create_session(self, globals, options=None):
        session = DemoSession(globals, options)
        self.sessions.append(session)
        return session


runtime = DemoRuntime()
qa = agent("question:string -> answer:string", {"runtime": {"language": "Python"}})
out = qa.test(runtime, "final()", {"question": "adapter"})
assert out["kind"] == "final", out
assert runtime.sessions[-1].closed

runner = agent("question:string -> answer:string", {"runtime": {"language": "Python"}})
step = runner.execute_actor_step(runtime, "final()", {"question": "adapter"})
assert step["kind"] == "final", step
snapshot = runner.export_session_state()
runner.restore_session_state(snapshot)
timeout = runner.execute_actor_step(runtime, "timeout()", {"question": "adapter"})
assert timeout["error_category"] == "timeout", timeout
print("python-runtime-adapter-ok")
